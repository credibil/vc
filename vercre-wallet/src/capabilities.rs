pub mod delay;
pub mod signer;
pub mod store;
