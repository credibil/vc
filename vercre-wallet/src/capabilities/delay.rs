//! Delay Capability
//!
//! This capability allows the application to delay the dispatch of an event by a
//!  specified amount of time.

use crux_core::capability::{CapabilityContext, Operation};
use crux_core::macros::Capability;
use serde::{Deserialize, Serialize};

/// Operations supported (by the Delay capability).
#[allow(clippy::module_name_repetitions)]
#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, Eq)]
pub struct DelayOperation {
    /// Delay in milliseconds.
    pub delay_ms: u64,
}

/// Output.
impl Operation for DelayOperation {
    type Output = ();
}

/// Delay capability context.
#[derive(Capability)]
#[allow(clippy::module_name_repetitions)]
pub struct Delay<Ev> {
    context: CapabilityContext<DelayOperation, Ev>,
}

impl<Ev> Delay<Ev>
where
    Ev: 'static,
{
    /// Create a new Delay capability context.
    #[must_use]
    pub const fn new(context: CapabilityContext<DelayOperation, Ev>) -> Self {
        Self { context }
    }

    /// Dispatch the specified event after a timer delay.
    pub fn start(&self, ms: u64, event: Ev)
    where
        Ev: Send + 'static,
    {
        self.context.spawn({
            let ctx = self.context.clone();

            async move {
                ctx.request_from_shell(DelayOperation { delay_ms: ms }).await;
                ctx.update_app(event);
            }
        });
    }
}
