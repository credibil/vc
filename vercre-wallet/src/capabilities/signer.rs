//! # Signer
//!
//! The `Signer` capability is used to sign requests and responses using the wallet's
//! signing key.
//!
//! Signing operations occur during:
//!
//! 1. Credential issuance — the wallet signs a Proof (JWT or CWT) as proof of possesion
//!    of key material. That is, proof that the wallet controls the private key for the
//!    DID the issued credential will be bound to.
//!
//! 2. Credential presentation — the wallet signs a Proof eincluded in the presentation
//!    response as proof of possession of key material. That is, proof that the wallet
//!    controls the private key for the DID the presentation response is bound to.

use std::fmt::Display;

use base64ct::{Base64UrlUnpadded, Encoding};
use crux_core::capability::{CapabilityContext, Operation};
use crux_core::macros::Capability;
use serde::{Deserialize, Serialize};
use thiserror::Error;
use vercre_core::jwt::Jwt;

/// Errors that can be returned by the Signer capability.
#[derive(Clone, Debug, Deserialize, Error, PartialEq, Eq)]
pub enum Error {
    /// The request was invalid.
    #[error("invalid signing request {0}")]
    InvalidRequest(String),
}

// manually implement serde::Serialize
impl Serialize for Error {
    fn serialize<S>(&self, serializer: S) -> core::result::Result<S::Ok, S::Error>
    where
        S: serde::ser::Serializer,
    {
        serializer.serialize_str(self.to_string().as_ref())
    }
}

impl From<serde_json::Error> for Error {
    fn from(err: serde_json::Error) -> Self {
        Self::InvalidRequest(err.to_string())
    }
}

/// Result type for the `Signer` capability.
pub type Result<T> = std::result::Result<T, Error>;

/// Operations supported (by the `Signer` capability).
#[derive(Clone, Serialize, Deserialize, Debug, PartialEq, Eq)]
#[allow(clippy::module_name_repetitions)]
pub enum SignerRequest {
    /// Sign the provided message
    Sign(Vec<u8>),

    /// Verification information
    Verification,
}

/// `SignerResponse` represents the output expected from any implementer of the
/// Signer capability.
#[derive(Clone, Serialize, Deserialize, Debug, PartialEq, Eq)]
#[allow(clippy::module_name_repetitions)]
pub enum SignerResponse {
    /// The signing operation returned the signature without issue.
    Signature(Vec<u8>),

    /// Signer's verification information.
    Verification {
        /// The algorithm used for signing.
        alg: String,

        /// The key identifier.
        kid: String,
    },

    /// The signing operation returned with the specified error.
    Err(String),
}

/// Operation provides a Crux wrapper for the `SignerRequest` consumed by the
/// Shell.
///
/// The Output type allows us to specify the expected response type from the
/// Shell.
impl Operation for SignerRequest {
    type Output = SignerResponse;
}

/// The Signer type used to implement the capability.
#[derive(Capability)]
pub struct Signer<Ev> {
    context: CapabilityContext<SignerRequest, Ev>,
}

impl<Ev> Signer<Ev>
where
    Ev: 'static,
{
    /// Create a new Signer capability context.
    #[must_use]
    pub const fn new(context: CapabilityContext<SignerRequest, Ev>) -> Self {
        Self { context }
    }

    /// Sign provided message using the vercre-wallet's signing key. Returns the signed output.
    ///
    /// Dispatches the 'callback' event specified in the capability request.
    ///
    /// # Panics
    ///
    /// Panics if the response is an `Err` variant.
    pub fn sign<T, F>(&self, jwt: &Jwt<T>, make_event: F)
    where
        T: Serialize + Clone,
        Jwt<T>: Display,
        F: Fn(Result<String>) -> Ev + Send + Sync + 'static,
    {
        self.context.spawn({
            let ctx = self.context.clone();
            let msg = jwt.to_string();

            async move {
                let request = SignerRequest::Sign(msg.clone().into_bytes());

                match ctx.request_from_shell(request).await {
                    SignerResponse::Signature(sig) => {
                        let sig_enc = Base64UrlUnpadded::encode_string(&sig);
                        let signed = format!("{msg}.{sig_enc}");
                        ctx.update_app(make_event(Ok(signed)));
                    }
                    SignerResponse::Err(err) => {
                        ctx.update_app(make_event(Err(Error::InvalidRequest(err))));
                    }
                    SignerResponse::Verification { .. } => {
                        panic!("Verification is not a valid response");
                    }
                }
            }
        });
    }

    /// Get the signer's verification information.
    ///
    /// Dispatches the 'callback' event specified in the capability request.
    ///
    /// # Panics
    ///
    /// Panics if the response is a `Signature` variant.
    pub fn verification<F>(&self, make_event: F)
    where
        F: Fn(Result<(String, String)>) -> Ev + Send + Sync + 'static,
    {
        self.context.spawn({
            let ctx = self.context.clone();

            async move {
                let request = SignerRequest::Verification;

                match ctx.request_from_shell(request).await {
                    SignerResponse::Verification { alg, kid } => {
                        ctx.update_app(make_event(Ok((alg, kid))));
                    }
                    SignerResponse::Err(err) => {
                        ctx.update_app(make_event(Err(Error::InvalidRequest(err))));
                    }
                    SignerResponse::Signature(_) => panic!("Signature is not a valid response"),
                }
            }
        });
    }
}
