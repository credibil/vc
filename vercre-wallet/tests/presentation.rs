mod verifier;

use assert_let_bind::assert_let;
use crux_core::testing::AppTester;
//
use crux_http::protocol::{HttpResponse, HttpResult};
use crux_http::testing::ResponseBuilder;
// use http_types::Body;
use insta::assert_yaml_snapshot as assert_snapshot;
// use insta::internals::Redaction;
use lazy_static::lazy_static;
use serde_json::{json, Value};
use test_utils::wallet;
use vercre_core::vp::{InvokeResponse, RequestObjectResponse, ResponseRequest};
use vercre_wallet::capabilities::signer::{SignerRequest, SignerResponse};
use vercre_wallet::capabilities::store::{StoreEntry, StoreRequest, StoreResponse};
use vercre_wallet::credential::Credential;
use vercre_wallet::presentation::*;

// Test that a 'cross-device' authorization (presentation) request event results in
// the requested VC's being presented.
#[tokio::test]
async fn cross_device_flow() {
    let verifier = verifier::new();

    // instantiate our app via the test harness
    let app = AppTester::<App, _>::default();
    let mut model = Model::default();

    // generate verifier request to 'send' to the app
    let body = json!({
        "purpose": "To verify employment",
        "input_descriptors": [{
            "id": "employment",
            "constraints": {
                "fields": [{
                    "path":["$.type"],
                    "filter": {
                        "type": "string",
                        "const": "EmployeeIDCredential"
                    }
                }]
            }
        }],
        "device_flow": "CrossDevice"
    });

    let resp = verifier.post("/invoke").json(&body).expect_success().await.json::<InvokeResponse>();
    let request_uri = resp.request_uri.expect("request uri");

    // ------------------------------------------------
    // Event::NewRequest
    //   1. creates a StoreRequest to find matching credentials
    //   2. emits Event::Matched
    // ------------------------------------------------
    let mut update = app.update(Event::Requested(request_uri.clone()), &mut model);
    assert_snapshot!("auth-request", app.view(&model));

    // verify the expected StoreRequest was emitted
    assert_let!(Effect::Http(request), &mut update.effects[0]);
    assert_eq!(request.operation.url, request_uri);

    let resp = verifier.get(&request_uri).expect_success().await;

    let req_obj_resp: RequestObjectResponse = resp.json();
    let http_resp = HttpResponse::ok().body(resp.into_bytes()).build();
    let update = app.resolve(request, HttpResult::Ok(http_resp)).expect("an update");

    // check the app emitted an (internal) event to update the model
    let response = ResponseBuilder::ok().body(req_obj_resp.clone()).build();
    assert_eq!(update.events, vec![Event::Fetched(Ok(response.clone()))]);

    // ------------------------------------------------
    // Event::Fetched
    //   1. creates a StoreRequest to find matching credentials
    //   2. emits Event::Matched
    // ------------------------------------------------
    let mut update = app.update(Event::Fetched(Ok(response)), &mut model);
    assert_snapshot!("set-request", app.view(&model));

    // verify the expected StoreRequest was emitted
    assert_let!(Effect::Store(request), &mut update.effects[0]);
    assert_eq!(StoreRequest::List, request.operation);

    // simulate the StoreResponse
    let matched: Credential =
        serde_json::from_value(APP_CREDENTIAL.to_owned()).expect("should deserialize");
    let matches = vec![matched];
    let results = matches
        .iter()
        .map(|m| StoreEntry(serde_json::to_vec(m).expect("should serialize")))
        .collect();
    let response = StoreResponse::List(results);
    let update = app.resolve(request, response).expect("an update");

    // verify the app emitted a Matched event
    let expected = Event::Matched(Ok(matches.clone()));
    assert_eq!(update.events[0], expected);

    // ------------------------------------------------
    // Event::Matched
    //   1. updates the model with the specified credentials
    // ------------------------------------------------
    app.update(Event::Matched(Ok(matches.clone())), &mut model);
    assert_snapshot!("set-credentials", app.view(&model), {
        ".credentials" => insta::sorted_redaction(),
        ".credentials[]" => insta::sorted_redaction(),
        ".credentials[].metadata.credential_definition.credentialSubject" => insta::sorted_redaction()
    });

    // ------------------------------------------------
    // Event::Authorize
    //   1. updates the model status to Authorized
    //   2. creates a SignerRequest get the signer's verification method
    //   3. emits a Proof event
    // ------------------------------------------------
    let mut update = app.update(Event::Authorized, &mut model);
    assert_snapshot!("authorize", app.view(&model), {
        ".credentials" => insta::sorted_redaction(),
        ".credentials[]" => insta::sorted_redaction(),
        ".credentials[].metadata.credential_definition.credentialSubject" => insta::sorted_redaction()
    });
    assert_let!(Effect::Signer(request), &mut update.effects[0]);
    assert_let!(SignerRequest::Verification, &request.operation);

    // simulate the SignerResponse
    let response = SignerResponse::Verification {
        alg: wallet::alg(),
        kid: wallet::kid(),
    };
    let update = app.resolve(request, response).expect("an update");

    // verify the app emitted a Proof event
    assert_let!(Event::Proof(Ok((alg, kid))), &update.events[0]);

    // ------------------------------------------------
    // Event::Proof
    //   1. updates the model with a submission
    //   2. creates a SignerRequest to sign the vp_token
    //   3. emits a Signed event on resolution of request
    // ------------------------------------------------
    let mut update = app.update(Event::Proof(Ok((alg.clone(), kid.clone()))), &mut model);
    assert_snapshot!("set-verification", app.view(&model), {
        ".credentials" => insta::sorted_redaction(),
        ".credentials[]" => insta::sorted_redaction(),
        ".credentials[].metadata.credential_definition.credentialSubject" => insta::sorted_redaction()
    });
    assert_let!(Effect::Signer(request), &mut update.effects[0]);
    assert_let!(SignerRequest::Sign(msg), &request.operation);

    // simulate the SignerResponse
    let signature = wallet::sign(msg);
    let response = SignerResponse::Signature(signature);
    let update = app.resolve(request, response).expect("an update");

    // verify the app emitted a Signed event
    assert_let!(Event::Signed(Ok(signed)), &update.events[0]);

    // ------------------------------------------------
    // Event::Signed
    //   1. creates a SignerRequest to sign the direct post jwt
    //   2. emits a JwtSigned event on resolution of request
    // ------------------------------------------------
    let mut update = app.update(Event::Signed(Ok(signed.to_string())), &mut model);
    assert_let!(Effect::Http(request), &mut update.effects[0]);
    // assert_eq!(request.operation.url, model.request.unwrap().response_uri.unwrap());

    // simulate the HttpResponse from our mock vercre-vp server
    let body = request.operation.body.clone();
    let form = serde_urlencoded::from_bytes::<ResponseRequest>(&body).expect("should deserialize");
    let res = verifier.post("/post").form(&form).expect_success().await;

    let resp: Value = res.json();
    let http_resp = HttpResponse::ok().body(res.into_bytes()).build();
    let update = app.resolve(request, HttpResult::Ok(http_resp)).expect("an update");

    // verify the expected Submitted event was emitted
    let response = ResponseBuilder::ok().body(resp.clone()).build();
    assert_eq!(update.events[0], Event::Submitted(Ok(response.clone())));

    // ------------------------------------------------
    // Event::VerifierResponse
    //   1. updates the model
    // ------------------------------------------------
    app.update(Event::Submitted(Ok(response)), &mut model);
    assert_snapshot!("verifier-response", app.view(&model));
}

lazy_static! {
    static ref APP_CREDENTIAL: Value = json!({
        "id":"61b9a42e-3dbb-4657-9a24-924a9586f233",
        "issuer":"http://localhost:8080",
        "metadata": {
            "format": "jwt_vc_json",
            "display":[{
                "name":"Employee ID",
                "description":"Vercre employee ID credential",
                "locale":"en-NZ",
                "logo":{"alt_text":"Vercre Logo","uri":"http://vercre.io/logo.png"},
                "text_color":"#ffffff",
                "background_color":"#12107c"
            }],
            "credential_definition": {
                "@context":[
                    "https://www.w3.org/2018/credentials/v1",
                    "https://www.w3.org/2018/credentials/examples/v1",
                ],
                "type":[
                    "VerifiableCredential",
                    "EmployeeIDCredential",
                ],
                "credentialSubject":{
                    "email":{
                        "display":[{"locale":"en-NZ","name":"Email"}],
                        "mandatory":true,"value_type":"string"
                    },
                    "familyName":{
                        "display":[{"locale":"en-NZ","name":"Family name"}],
                        "mandatory":true,
                        "value_type":"string"
                    },
                    "givenName":{
                        "display":[{"locale":"en-NZ","name":"Given name"}],
                        "mandatory":true,
                        "value_type":"string"
                    }
                }
            }
        },
        "vc":{
            "@context": [
                "https://www.w3.org/2018/credentials/v1",
                "http://vercre.io/credentials/v1"
            ],
            "id": "EmployeeID_JWT",
            "type": [
                "VerifiableCredential",
                "EmployeeIDCredential"
            ],
            "issuer": "http://vercre.io",
            "issuanceDate": "2023-11-20T23:21:55.488659Z",
            "credentialSubject": {
                "id": "did:ion:EiDyOQbbZAa3aiRzeCkV7LOx3SERjjH93EXoIM3UoN4oWg:eyJkZWx0YSI6eyJwYXRjaGVzIjpbeyJhY3Rpb24iOiJyZXBsYWNlIiwiZG9jdW1lbnQiOnsicHVibGljS2V5cyI6W3siaWQiOiJwdWJsaWNLZXlNb2RlbDFJZCIsInB1YmxpY0tleUp3ayI6eyJjcnYiOiJzZWNwMjU2azEiLCJrdHkiOiJFQyIsIngiOiJ0WFNLQl9ydWJYUzdzQ2pYcXVwVkpFelRjVzNNc2ptRXZxMVlwWG45NlpnIiwieSI6ImRPaWNYcWJqRnhvR0otSzAtR0oxa0hZSnFpY19EX09NdVV3a1E3T2w2bmsifSwicHVycG9zZXMiOlsiYXV0aGVudGljYXRpb24iLCJrZXlBZ3JlZW1lbnQiXSwidHlwZSI6IkVjZHNhU2VjcDI1NmsxVmVyaWZpY2F0aW9uS2V5MjAxOSJ9XSwic2VydmljZXMiOlt7ImlkIjoic2VydmljZTFJZCIsInNlcnZpY2VFbmRwb2ludCI6Imh0dHA6Ly93d3cuc2VydmljZTEuY29tIiwidHlwZSI6InNlcnZpY2UxVHlwZSJ9XX19XSwidXBkYXRlQ29tbWl0bWVudCI6IkVpREtJa3dxTzY5SVBHM3BPbEhrZGI4Nm5ZdDBhTnhTSFp1MnItYmhFem5qZEEifSwic3VmZml4RGF0YSI6eyJkZWx0YUhhc2giOiJFaUNmRFdSbllsY0Q5RUdBM2RfNVoxQUh1LWlZcU1iSjluZmlxZHo1UzhWRGJnIiwicmVjb3ZlcnlDb21taXRtZW50IjoiRWlCZk9aZE10VTZPQnc4UGs4NzlRdFotMkotOUZiYmpTWnlvYUFfYnFENHpoQSJ9fQ"
            },
            "proof": {
                "type": "",
                "cryptosuite": "EcdsaSecp256k1VerificationKey2019",
                "proofPurpose": "",
                "verificationMethod": "did:ion:EiDyOQbbZAa3aiRzeCkV7LOx3SERjjH93EXoIM3UoN4oWg:eyJkZWx0YSI6eyJwYXRjaGVzIjpbeyJhY3Rpb24iOiJyZXBsYWNlIiwiZG9jdW1lbnQiOnsicHVibGljS2V5cyI6W3siaWQiOiJwdWJsaWNLZXlNb2RlbDFJZCIsInB1YmxpY0tleUp3ayI6eyJjcnYiOiJzZWNwMjU2azEiLCJrdHkiOiJFQyIsIngiOiJ0WFNLQl9ydWJYUzdzQ2pYcXVwVkpFelRjVzNNc2ptRXZxMVlwWG45NlpnIiwieSI6ImRPaWNYcWJqRnhvR0otSzAtR0oxa0hZSnFpY19EX09NdVV3a1E3T2w2bmsifSwicHVycG9zZXMiOlsiYXV0aGVudGljYXRpb24iLCJrZXlBZ3JlZW1lbnQiXSwidHlwZSI6IkVjZHNhU2VjcDI1NmsxVmVyaWZpY2F0aW9uS2V5MjAxOSJ9XSwic2VydmljZXMiOlt7ImlkIjoic2VydmljZTFJZCIsInNlcnZpY2VFbmRwb2ludCI6Imh0dHA6Ly93d3cuc2VydmljZTEuY29tIiwidHlwZSI6InNlcnZpY2UxVHlwZSJ9XX19XSwidXBkYXRlQ29tbWl0bWVudCI6IkVpREtJa3dxTzY5SVBHM3BPbEhrZGI4Nm5ZdDBhTnhTSFp1MnItYmhFem5qZEEifSwic3VmZml4RGF0YSI6eyJkZWx0YUhhc2giOiJFaUNmRFdSbllsY0Q5RUdBM2RfNVoxQUh1LWlZcU1iSjluZmlxZHo1UzhWRGJnIiwicmVjb3ZlcnlDb21taXRtZW50IjoiRWlCZk9aZE10VTZPQnc4UGs4NzlRdFotMkotOUZiYmpTWnlvYUFfYnFENHpoQSJ9fQ#publicKeyModel1Id",
                "proofValue": ""
            }
        },
        "issued":"eyJhbGciOiJFUzI1NksiLCJraWQiOiJkaWQ6aW9uOkVpRHlPUWJiWkFhM2FpUnplQ2tWN0xPeDNTRVJqakg5M0VYb0lNM1VvTjRvV2c6ZXlKa1pXeDBZU0k2ZXlKd1lYUmphR1Z6SWpwYmV5SmhZM1JwYjI0aU9pSnlaWEJzWVdObElpd2laRzlqZFcxbGJuUWlPbnNpY0hWaWJHbGpTMlY1Y3lJNlczc2lhV1FpT2lKd2RXSnNhV05MWlhsTmIyUmxiREZKWkNJc0luQjFZbXhwWTB0bGVVcDNheUk2ZXlKamNuWWlPaUp6WldOd01qVTJhekVpTENKcmRIa2lPaUpGUXlJc0luZ2lPaUowV0ZOTFFsOXlkV0pZVXpkelEycFljWFZ3VmtwRmVsUmpWek5OYzJwdFJYWnhNVmx3V0c0NU5scG5JaXdpZVNJNkltUlBhV05ZY1dKcVJuaHZSMG90U3pBdFIwb3hhMGhaU25GcFkxOUVYMDlOZFZWM2ExRTNUMncyYm1zaWZTd2ljSFZ5Y0c5elpYTWlPbHNpWVhWMGFHVnVkR2xqWVhScGIyNGlMQ0pyWlhsQlozSmxaVzFsYm5RaVhTd2lkSGx3WlNJNklrVmpaSE5oVTJWamNESTFObXN4Vm1WeWFXWnBZMkYwYVc5dVMyVjVNakF4T1NKOVhTd2ljMlZ5ZG1salpYTWlPbHQ3SW1sa0lqb2ljMlZ5ZG1salpURkpaQ0lzSW5ObGNuWnBZMlZGYm1Sd2IybHVkQ0k2SW1oMGRIQTZMeTkzZDNjdWMyVnlkbWxqWlRFdVkyOXRJaXdpZEhsd1pTSTZJbk5sY25acFkyVXhWSGx3WlNKOVhYMTlYU3dpZFhCa1lYUmxRMjl0YldsMGJXVnVkQ0k2SWtWcFJFdEphM2R4VHpZNVNWQkhNM0JQYkVoclpHSTRObTVaZERCaFRuaFRTRnAxTW5JdFltaEZlbTVxWkVFaWZTd2ljM1ZtWm1sNFJHRjBZU0k2ZXlKa1pXeDBZVWhoYzJnaU9pSkZhVU5tUkZkU2JsbHNZMFE1UlVkQk0yUmZOVm94UVVoMUxXbFpjVTFpU2psdVptbHhaSG8xVXpoV1JHSm5JaXdpY21WamIzWmxjbmxEYjIxdGFYUnRaVzUwSWpvaVJXbENaazlhWkUxMFZUWlBRbmM0VUdzNE56bFJkRm90TWtvdE9VWmlZbXBUV25sdllVRmZZbkZFTkhwb1FTSjlmUSNwdWJsaWNLZXlNb2RlbDFJZCIsInR5cCI6Imp3dCJ9.eyJzdWIiOiJkaWQ6aW9uOkVpRHlPUWJiWkFhM2FpUnplQ2tWN0xPeDNTRVJqakg5M0VYb0lNM1VvTjRvV2c6ZXlKa1pXeDBZU0k2ZXlKd1lYUmphR1Z6SWpwYmV5SmhZM1JwYjI0aU9pSnlaWEJzWVdObElpd2laRzlqZFcxbGJuUWlPbnNpY0hWaWJHbGpTMlY1Y3lJNlczc2lhV1FpT2lKd2RXSnNhV05MWlhsTmIyUmxiREZKWkNJc0luQjFZbXhwWTB0bGVVcDNheUk2ZXlKamNuWWlPaUp6WldOd01qVTJhekVpTENKcmRIa2lPaUpGUXlJc0luZ2lPaUowV0ZOTFFsOXlkV0pZVXpkelEycFljWFZ3VmtwRmVsUmpWek5OYzJwdFJYWnhNVmx3V0c0NU5scG5JaXdpZVNJNkltUlBhV05ZY1dKcVJuaHZSMG90U3pBdFIwb3hhMGhaU25GcFkxOUVYMDlOZFZWM2ExRTNUMncyYm1zaWZTd2ljSFZ5Y0c5elpYTWlPbHNpWVhWMGFHVnVkR2xqWVhScGIyNGlMQ0pyWlhsQlozSmxaVzFsYm5RaVhTd2lkSGx3WlNJNklrVmpaSE5oVTJWamNESTFObXN4Vm1WeWFXWnBZMkYwYVc5dVMyVjVNakF4T1NKOVhTd2ljMlZ5ZG1salpYTWlPbHQ3SW1sa0lqb2ljMlZ5ZG1salpURkpaQ0lzSW5ObGNuWnBZMlZGYm1Sd2IybHVkQ0k2SW1oMGRIQTZMeTkzZDNjdWMyVnlkbWxqWlRFdVkyOXRJaXdpZEhsd1pTSTZJbk5sY25acFkyVXhWSGx3WlNKOVhYMTlYU3dpZFhCa1lYUmxRMjl0YldsMGJXVnVkQ0k2SWtWcFJFdEphM2R4VHpZNVNWQkhNM0JQYkVoclpHSTRObTVaZERCaFRuaFRTRnAxTW5JdFltaEZlbTVxWkVFaWZTd2ljM1ZtWm1sNFJHRjBZU0k2ZXlKa1pXeDBZVWhoYzJnaU9pSkZhVU5tUkZkU2JsbHNZMFE1UlVkQk0yUmZOVm94UVVoMUxXbFpjVTFpU2psdVptbHhaSG8xVXpoV1JHSm5JaXdpY21WamIzWmxjbmxEYjIxdGFYUnRaVzUwSWpvaVJXbENaazlhWkUxMFZUWlBRbmM0VUdzNE56bFJkRm90TWtvdE9VWmlZbXBUV25sdllVRmZZbkZFTkhwb1FTSjlmUSIsImp0aSI6IkVtcGxveWVlSURfSldUIiwiaXNzIjoiaHR0cDovL2NyZWRpYmlsLmlvIiwibmJmIjoxNzAwNTIyNTE1LCJpYXQiOjE3MDA1MjI1MTUsImV4cCI6bnVsbCwidmMiOnsiQGNvbnRleHQiOlsiaHR0cHM6Ly93d3cudzMub3JnLzIwMTgvY3JlZGVudGlhbHMvdjEiLCJodHRwOi8vY3JlZGliaWwuaW8vY3JlZGVudGlhbHMvdjEiXSwiaWQiOiJFbXBsb3llZUlEX0pXVCIsInR5cGUiOlsiVmVyaWZpYWJsZUNyZWRlbnRpYWwiLCJFbXBsb3llZUlEQ3JlZGVudGlhbCJdLCJpc3N1ZXIiOiJodHRwOi8vY3JlZGliaWwuaW8iLCJpc3N1YW5jZURhdGUiOiIyMDIzLTExLTIwVDIzOjIxOjU1LjQ4ODY1OVoiLCJjcmVkZW50aWFsU3ViamVjdCI6eyJpZCI6ImRpZDppb246RWlEeU9RYmJaQWEzYWlSemVDa1Y3TE94M1NFUmpqSDkzRVhvSU0zVW9ONG9XZzpleUprWld4MFlTSTZleUp3WVhSamFHVnpJanBiZXlKaFkzUnBiMjRpT2lKeVpYQnNZV05sSWl3aVpHOWpkVzFsYm5RaU9uc2ljSFZpYkdsalMyVjVjeUk2VzNzaWFXUWlPaUp3ZFdKc2FXTkxaWGxOYjJSbGJERkpaQ0lzSW5CMVlteHBZMHRsZVVwM2F5STZleUpqY25ZaU9pSnpaV053TWpVMmF6RWlMQ0pyZEhraU9pSkZReUlzSW5naU9pSjBXRk5MUWw5eWRXSllVemR6UTJwWWNYVndWa3BGZWxSalZ6Tk5jMnB0UlhaeE1WbHdXRzQ1Tmxwbklpd2llU0k2SW1SUGFXTlljV0pxUm5odlIwb3RTekF0UjBveGEwaFpTbkZwWTE5RVgwOU5kVlYzYTFFM1QydzJibXNpZlN3aWNIVnljRzl6WlhNaU9sc2lZWFYwYUdWdWRHbGpZWFJwYjI0aUxDSnJaWGxCWjNKbFpXMWxiblFpWFN3aWRIbHdaU0k2SWtWalpITmhVMlZqY0RJMU5tc3hWbVZ5YVdacFkyRjBhVzl1UzJWNU1qQXhPU0o5WFN3aWMyVnlkbWxqWlhNaU9sdDdJbWxrSWpvaWMyVnlkbWxqWlRGSlpDSXNJbk5sY25acFkyVkZibVJ3YjJsdWRDSTZJbWgwZEhBNkx5OTNkM2N1YzJWeWRtbGpaVEV1WTI5dElpd2lkSGx3WlNJNkluTmxjblpwWTJVeFZIbHdaU0o5WFgxOVhTd2lkWEJrWVhSbFEyOXRiV2wwYldWdWRDSTZJa1ZwUkV0SmEzZHhUelk1U1ZCSE0zQlBiRWhyWkdJNE5tNVpkREJoVG5oVFNGcDFNbkl0WW1oRmVtNXFaRUVpZlN3aWMzVm1abWw0UkdGMFlTSTZleUprWld4MFlVaGhjMmdpT2lKRmFVTm1SRmRTYmxsc1kwUTVSVWRCTTJSZk5Wb3hRVWgxTFdsWmNVMWlTamx1Wm1seFpIbzFVemhXUkdKbklpd2ljbVZqYjNabGNubERiMjF0YVhSdFpXNTBJam9pUldsQ1prOWFaRTEwVlRaUFFuYzRVR3M0TnpsUmRGb3RNa290T1VaaVltcFRXbmx2WVVGZlluRkVOSHBvUVNKOWZRIn0sInByb29mIjp7InR5cGUiOiIiLCJjcnlwdG9zdWl0ZSI6IkVjZHNhU2VjcDI1NmsxVmVyaWZpY2F0aW9uS2V5MjAxOSIsInByb29mUHVycG9zZSI6IiIsInZlcmlmaWNhdGlvbk1ldGhvZCI6ImRpZDppb246RWlEeU9RYmJaQWEzYWlSemVDa1Y3TE94M1NFUmpqSDkzRVhvSU0zVW9ONG9XZzpleUprWld4MFlTSTZleUp3WVhSamFHVnpJanBiZXlKaFkzUnBiMjRpT2lKeVpYQnNZV05sSWl3aVpHOWpkVzFsYm5RaU9uc2ljSFZpYkdsalMyVjVjeUk2VzNzaWFXUWlPaUp3ZFdKc2FXTkxaWGxOYjJSbGJERkpaQ0lzSW5CMVlteHBZMHRsZVVwM2F5STZleUpqY25ZaU9pSnpaV053TWpVMmF6RWlMQ0pyZEhraU9pSkZReUlzSW5naU9pSjBXRk5MUWw5eWRXSllVemR6UTJwWWNYVndWa3BGZWxSalZ6Tk5jMnB0UlhaeE1WbHdXRzQ1Tmxwbklpd2llU0k2SW1SUGFXTlljV0pxUm5odlIwb3RTekF0UjBveGEwaFpTbkZwWTE5RVgwOU5kVlYzYTFFM1QydzJibXNpZlN3aWNIVnljRzl6WlhNaU9sc2lZWFYwYUdWdWRHbGpZWFJwYjI0aUxDSnJaWGxCWjNKbFpXMWxiblFpWFN3aWRIbHdaU0k2SWtWalpITmhVMlZqY0RJMU5tc3hWbVZ5YVdacFkyRjBhVzl1UzJWNU1qQXhPU0o5WFN3aWMyVnlkbWxqWlhNaU9sdDdJbWxrSWpvaWMyVnlkbWxqWlRGSlpDSXNJbk5sY25acFkyVkZibVJ3YjJsdWRDSTZJbWgwZEhBNkx5OTNkM2N1YzJWeWRtbGpaVEV1WTI5dElpd2lkSGx3WlNJNkluTmxjblpwWTJVeFZIbHdaU0o5WFgxOVhTd2lkWEJrWVhSbFEyOXRiV2wwYldWdWRDSTZJa1ZwUkV0SmEzZHhUelk1U1ZCSE0zQlBiRWhyWkdJNE5tNVpkREJoVG5oVFNGcDFNbkl0WW1oRmVtNXFaRUVpZlN3aWMzVm1abWw0UkdGMFlTSTZleUprWld4MFlVaGhjMmdpT2lKRmFVTm1SRmRTYmxsc1kwUTVSVWRCTTJSZk5Wb3hRVWgxTFdsWmNVMWlTamx1Wm1seFpIbzFVemhXUkdKbklpd2ljbVZqYjNabGNubERiMjF0YVhSdFpXNTBJam9pUldsQ1prOWFaRTEwVlRaUFFuYzRVR3M0TnpsUmRGb3RNa290T1VaaVltcFRXbmx2WVVGZlluRkVOSHBvUVNKOWZRI3B1YmxpY0tleU1vZGVsMUlkIiwicHJvb2ZWYWx1ZSI6IiJ9fX0.yUsvBJDMk5rS7BjGlOT4TwUeI4IczC5RihwNSm4ErRgd8CfSdf0aEIzMGcHxxYNVaMHPV0yzM8VgC0jLsv14aQ"});
}
