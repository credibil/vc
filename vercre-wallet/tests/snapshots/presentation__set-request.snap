---
source: vercre-wallet/tests/presentation.rs
assertion_line: 86
expression: app.view(&model)
---
credentials: []
status: Requested

